### PR TITLE
Remove +1 assumption for input versions in $import

### PR DESCRIPTION
**Pull request by SergeyGaluzo**
There is demand to honor arbitrary non sequential resource versions provided for $import. Currently versions are assumed to be sequential ints. This PR enables any int version.

### CODE REVIEW

File: test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Import/ImportTests.cs [hunk 1]
```diff
@@ -92,6 +92,39 @@ public async Task GivenIncrementalLoad_MultipleInputVersionsOutOfOrderSomeNotExp
             Assert.Equal(GetLastUpdated("2002"), result.Resource.Meta.LastUpdated);
         }
 
+        [Fact]
```
Comment (abiisnn) [inline test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Import/ImportTests.cs:95]: Can we add a test where we provide non sequential resource versions?

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs [hunk 2]
```diff
@@ -164,12 +164,12 @@ internal class SqlServerFhirDataStore : IFhirDataStore, IProvideCapability
             (var transactionId, var minSequenceId) = await StoreClient.MergeResourcesBeginTransactionAsync(resources.Count, cancellationToken);
 
             var index = 0;
-            var mergeWrappers = new List<MergeResourceWrapper>();
+            var mergeWrappersPlus = new List<(MergeResourceWrapper Wrapper, bool KeepVersion, int ResourceVersion, int? ExistingVersion)>();
```
[the remark this answers is not in the document]
Comment (SergeyGaluzo) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs:167]: Plus indicates that this variable contains not only merge wrappers but other data points too. There are 3 other data points.

File: test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Import/ImportTests.cs (hunk 1 above)
Comment (SergeyGaluzo) [reply test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Import/ImportTests.cs:95]: Not sure what you mean. Input order of records is immaterial as all are sorted internally.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs:167]: mergeWrappersAndVersions :)

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs [hunk 3]
```diff
@@ -310,12 +310,34 @@ internal class SqlServerFhirDataStore : IFhirDataStore, IProvideCapability
...
-                mergeWrappers.Add(new MergeResourceWrapper(resource, resourceExt.KeepHistory, hasVersionToCompare));
+                mergeWrappersPlus.Add((new MergeResourceWrapper(resource, resourceExt.KeepHistory, hasVersionToCompare), resourceExt.KeepVersion, int.Parse(resource.Version), existingVersion));
                 index++;
                 results.Add(identifier, new DataStoreOperationOutcome(new UpsertOutcome(resource, resource.Version == InitialVersion ? SaveOutcomeType.Created : SaveOutcomeType.Updated)));
             }
 
-            if (mergeWrappers.Count > 0) // Do not call DB with empty input
+            // Resources with input versions (keepVersion=true) might not have hasVersionToCompare set. Fix it here.
+            // Resources with keepVersion=true must be in separate call, and not mixed with keepVersion=false ones.
+            // Sort them in groups by resource id and order by version.
```
[the remark this answers is not in the document]
Comment (SergeyGaluzo) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs:320]: I was not able to find a clear way to apply required logic only using Linq. So I did what I did. I hope it is obvious, hence simple.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs [hunk 4]
```diff
@@ -310,12 +310,34 @@ internal class SqlServerFhirDataStore : IFhirDataStore, IProvideCapability
...
+            // In each group find the smallest version higher then existing
+            prevResourceId = string.Empty;
+            var notSetInResoureGroup = false;
+            foreach (var mergeWrapperPlus in mergeWrappersPlus.Where(_ => _.KeepVersion && _.ExistingVersion != 0).OrderBy(_ => _.Wrapper.ResourceWrapper.ResourceId).ThenBy(_ => _.ResourceVersion))
```
[the remark this answers is not in the document]
Comment (SergeyGaluzo) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs:324]: :-) I like _ as it denotes that it is derived from the definition of enumerable, nothing else. In complex cases when there is more than one thing derived, I use meaningful variable names to differentiate between cases. BTW All code scanner are happy...

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (SergeyGaluzo) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs:167]: Fine.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs]: image

image

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs [hunk 5]
```diff
@@ -310,12 +310,34 @@ internal class SqlServerFhirDataStore : IFhirDataStore, IProvideCapability
...
+                }
+
+                prevResourceId = mergeWrapper.Wrapper.ResourceWrapper.ResourceId;
+
+                if (notSetInResoureGroup && mergeWrapper.ResourceVersion > mergeWrapper.ExistingVersion)
+                {
+                    mergeWrapper.Wrapper.HasVersionToCompare = true;
+                    notSetInResoureGroup = false;
+                }
+            }
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs:338]: The linq version would look like this:

```csharp
foreach (var wrapperWithVersionToCompare in mergeWrappersWithVersions
             .Where(x => x.KeepVersion && x.ExistingVersion != 0)
             .GroupBy(key => key.Wrapper.ResourceWrapper.ResourceId)
             .Select(group => group.MinBy(y => y.ResourceVersion))
             .Where(x => x.ResourceVersion > x.ExistingVersion))
{
    wrapperWithVersionToCompare.Wrapper.HasVersionToCompare = true;
}
```

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (SergeyGaluzo) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs]: When hovering over _, my VS shows correct type details.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs (hunk 5 above)
Comment (SergeyGaluzo) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs:338]: Cool. It is more difficult for me to understand, so I want to stick to my current version.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs]: It's also documented in the C# fundamentals: https://learn.microsoft.com/en-us/dotnet/csharp/fundamentals/functional/discards

image

[the remark this answers is not in the document]
Comment (SergeyGaluzo) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs]: Replaced _ by x.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs (hunk 5 above)
Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs]: Sergey, I understand you have code that functions as intended. In this case though, .net has a `GroupBy` function that we can leverage.

**Review by brendankowitz — APPROVED** [review]

**Pull request merged**